Improve script execution context usage in Trusted Types code

lukewarlow opened the pull request:
#### 76ec38e84a6401d83ff18fc5d0a24f71a062bf56
<pre>
Improve script execution context usage in Trusted Types code
<a href="https://bugs.webkit.org/show_bug.cgi?id=301625">https://bugs.webkit.org/show_bug.cgi?id=301625</a>

Reviewed by Darin Adler.

The existing usage of document() rather than document().contextDocument() can cause situations where
objects created within secondary documents (e.g. createHTMLDocument()) would not be protected by Trusted Types.

Test: imported/w3c/web-platform-tests/trusted-types/trusted-types-secondary-document.html
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/trusted-types-secondary-document-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/trusted-types-secondary-document.html: Added.
* Source/WebCore/dom/Attr.cpp:
(WebCore::Attr::setValue):
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::parseHTMLUnsafe):
(WebCore::Document::write):
(WebCore::Document::execCommand):
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::setAttribute):
(WebCore::Element::setAttributeNode):
(WebCore::Element::setAttributeNodeNS):
(WebCore::Element::setAttributeNS):
(WebCore::Element::setHTMLUnsafe):
(WebCore::Element::setOuterHTML):
(WebCore::Element::setInnerHTML):
(WebCore::Element::insertAdjacentHTML):
* Source/WebCore/dom/Range.cpp:
(WebCore::Range::createContextualFragment):
* Source/WebCore/dom/ShadowRoot.cpp:
(WebCore::ShadowRoot::setHTMLUnsafe):
(WebCore::ShadowRoot::setInnerHTML):
* Source/WebCore/html/HTMLIFrameElement.cpp:
(WebCore::HTMLIFrameElement::setSrcdoc):
* Source/WebCore/svg/properties/SVGAnimatedString.cpp:
(WebCore::SVGAnimatedString::setBaseVal):
* Source/WebCore/workers/shared/SharedWorker.cpp:
(WebCore::SharedWorker::create):
* Source/WebCore/xml/DOMParser.cpp:
(WebCore::DOMParser::parseFromString):

Canonical link: <a href="https://commits.webkit.org/302412@main">https://commits.webkit.org/302412@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7a29ebc892a3ed8b191a1fbd9d28de57c8ce471d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/128863 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/1120 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/39695 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/136246 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/80231 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/130734 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/1072 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/998 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/98103 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/66050 "") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/131810 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/812 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/115435 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/78715 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/742 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/33546 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/79525 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/109179 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/34040 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/138707 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/937 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/906 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/106641 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/990 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/111769 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/106452 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/761 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/30301 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/53393 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/20144 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/1005 "Built successfully") | [❌ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/64314 "Found 158 new failures in WebProcess/Extensions/API/Cocoa/WebExtensionAPIRuntimeCocoa.mm, WebProcess/Extensions/API/Cocoa/WebExtensionAPITabsCocoa.mm, WebProcess/Extensions/API/Cocoa/WebExtensionAPIStorageCocoa.mm, WebProcess/Extensions/API/Cocoa/WebExtensionAPIMenusCocoa.mm, workers/shared/SharedWorker.cpp, WebProcess/InjectedBundle/API/c/WKBundlePage.cpp, WebProcess/Extensions/API/Cocoa/WebExtensionAPIPortCocoa.mm, WebProcess/WebPage/RemoteLayerTree/RemoteLayerTreeDrawingArea.mm, WebProcess/Extensions/API/Cocoa/WebExtensionAPIStorageAreaCocoa.mm, WebProcess/Extensions/API/Cocoa/WebExtensionAPINamespaceCocoa.mm ...") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/850 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/906 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/941 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/xml/DOMParser.cpp:
```diff
@@ -46,7 +46,7 @@ Ref<DOMParser> DOMParser::create(Document& contextDocument)
 ExceptionOr<Ref<Document>> DOMParser::parseFromString(Variant<RefPtr<TrustedHTML>, String>&& string, const AtomString& contentType)
 {
-    auto stringValueHolder = trustedTypeCompliantString(*protectedContextDocument()->protectedScriptExecutionContext().get(), WTFMove(string), "DOMParser parseFromString"_s);
+    auto stringValueHolder = trustedTypeCompliantString(*protectedContextDocument().get(), WTFMove(string), "DOMParser parseFromString"_s);
```
Does it work without `get()`? I think the `*` operator works on `RefPtr` directly.
```suggestion
    auto stringValueHolder = trustedTypeCompliantString(*protectedContextDocument(), WTFMove(string), "DOMParser parseFromString"_s);
```

lukewarlow replied:
Good point.

lukewarlow commented on LayoutTests/imported/w3c/web-platform-tests/trusted-types/trusted-types-secondary-document.html:
```diff
@@ -0,0 +1,340 @@
+<!DOCTYPE html>
```
https://github.com/web-platform-tests/wpt/pull/55772

lukewarlow committed:
Improve script execution context usage in Trusted Types code
https://bugs.webkit.org/show_bug.cgi?id=301625

Reviewed by Darin Adler.

The existing usage of document() rather than document().contextDocument() can cause situations where
objects created within secondary documents (e.g. createHTMLDocument()) would not be protected by Trusted Types.

Test: imported/w3c/web-platform-tests/trusted-types/trusted-types-secondary-document.html
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/trusted-types-secondary-document-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/trusted-types-secondary-document.html: Added.
* Source/WebCore/dom/Attr.cpp:
(WebCore::Attr::setValue):
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::parseHTMLUnsafe):
(WebCore::Document::write):
(WebCore::Document::execCommand):
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::setAttribute):
(WebCore::Element::setAttributeNode):
(WebCore::Element::setAttributeNodeNS):
(WebCore::Element::setAttributeNS):
(WebCore::Element::setHTMLUnsafe):
(WebCore::Element::setOuterHTML):
(WebCore::Element::setInnerHTML):
(WebCore::Element::insertAdjacentHTML):
* Source/WebCore/dom/Range.cpp:
(WebCore::Range::createContextualFragment):
* Source/WebCore/dom/ShadowRoot.cpp:
(WebCore::ShadowRoot::setHTMLUnsafe):
(WebCore::ShadowRoot::setInnerHTML):
* Source/WebCore/html/HTMLIFrameElement.cpp:
(WebCore::HTMLIFrameElement::setSrcdoc):
* Source/WebCore/svg/properties/SVGAnimatedString.cpp:
(WebCore::SVGAnimatedString::setBaseVal):
* Source/WebCore/workers/shared/SharedWorker.cpp:
(WebCore::SharedWorker::create):
* Source/WebCore/xml/DOMParser.cpp:
(WebCore::DOMParser::parseFromString):

Canonical link: https://commits.webkit.org/302412@main

darinadler reviewed (approved):
Is there enough test coverage here? I hope that all the substantive changes are needed to fix some test case. In many of these cases it’s absolutely not clear why it needs to be the result of `contextDocument()` and not the document itself. Type checking doesn’t help since both are script execution contexts. So we need tests to help us know we have it all correct.
